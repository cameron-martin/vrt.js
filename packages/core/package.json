{
    "name": "@vrt.js/core",
    "version": "0.0.1",
    "main": "dist/index.js",
    "dependencies": {
        "combine-async-iterators": "^1.1.1",
        "resemblejs": "^3.2.0"
    }
}
