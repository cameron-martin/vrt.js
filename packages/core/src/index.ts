export { Screenshot, ScreenshotKey, Backend } from './Backend';
export { Browser, BrowserSession } from './Browser';
export { MatrixBackend, CompositeBackend } from './combinators';
export { Configuration } from './Configuration';
export { Report, ScreenshotReport, Reporter } from './Reporter';
export { default as run } from './run';
