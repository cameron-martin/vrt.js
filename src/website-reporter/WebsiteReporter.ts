import { Reporter, Report, ScreenshotReport } from '../core/Reporter';
import fs from 'fs-extra';
import path from 'path';
import ejs from 'ejs';

function createIdGenerator() {
  let id = 0;
  return () => ++id;
}

type ScreenshotKeys = 'before' | 'after' | 'diff';

type ScreenshotViewModel = {
  [K in keyof ScreenshotReport]: K extends ScreenshotKeys
    ? string
    : ScreenshotReport[K];
};

interface Config {
  outputDirectory: string;
}

export default class WebsiteReporter implements Reporter {
  private screenshotsDir = path.join(
    this.config.outputDirectory,
    'screenshots',
  );
  private reportUrl = path.join(this.config.outputDirectory, 'index.html');

  constructor(private readonly config: Config) {}

  async report(report: Report): Promise<void> {
    await fs.emptyDir(this.config.outputDirectory);
    await fs.mkdir(this.screenshotsDir);

    const generateId = createIdGenerator();

    const screenshots = await Promise.all(
      report.screenshots.map(screenshot =>
        this.getScreenshotViewModel(screenshot, generateId),
      ),
    );

    const result = await ejs.renderFile<string>(
      require.resolve('./templates/index.ejs'),
      {
        screenshots,
      },
    );

    await fs.writeFile(this.reportUrl, result, 'utf8');
  }

  async getScreenshotViewModel(
    screenshot: ScreenshotReport,
    generateId: () => number,
  ): Promise<ScreenshotViewModel> {
    const [before, after, diff] = await Promise.all(
      [screenshot.before, screenshot.after, screenshot.diff].map(
        async screenshot => {
          const id = generateId();
          const filePath = path.join(this.screenshotsDir, `${id}.png`);

          await fs.writeFile(filePath, screenshot);

          return path.relative(path.dirname(this.reportUrl), filePath);
        },
      ),
    );

    return {
      ...screenshot,
      before,
      after,
      diff,
    };
  }
}
