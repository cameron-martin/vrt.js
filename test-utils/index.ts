import fs from 'fs-extra';
import path from 'path';
import { Backend, Screenshot, Reporter, Report } from '@vrt.js/core';

export class MockBackend implements Backend {
  constructor(private readonly screenshots: Screenshot[]) {}
  async *getScreenshots() {
    for (const screenshot of this.screenshots) {
      yield screenshot;
    }
  }
}

export class MockReporter implements Reporter {
  private _report: Report | undefined;

  async report(report: Report): Promise<void> {
    if (!this._report) {
      this._report = report;
    } else {
      throw new Error('Reporter was called multiple times');
    }
  }

  getReport(): Report {
    if (this._report) {
      return this._report;
    } else {
      throw new Error('Reporter was not called yet');
    }
  }
}

export async function asyncIterableToArray<T>(
  iterable: AsyncIterable<T>,
): Promise<T[]> {
  const items: T[] = [];

  for await (const item of iterable) {
    items.push(item);
  }

  return items;
}

export function getExampleScreenshot(name: string): Promise<Buffer> {
  return fs.readFile(path.join(__dirname, 'screenshot-examples', name));
}
