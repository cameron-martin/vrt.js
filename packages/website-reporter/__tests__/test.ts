import WebsiteReporter from '../';
import tempy from 'tempy';
import { Report } from '@vrt.js/core';
import fs from 'fs-extra';
import path from 'path';
import { getExampleScreenshot } from '../../../test-utils';

it('matches snapshot', async () => {
  const outputDirectory = tempy.directory();

  const websiteReporter = new WebsiteReporter({
    outputDirectory,
  });

  const exampleScreenshot = await getExampleScreenshot('screenshot1.png');

  const report: Report = {
    screenshots: [
      {
        properties: { key: '/foo', browser: 'Chrome', viewportWidth: 1024 },
        before: exampleScreenshot,
        after: exampleScreenshot,
        diff: exampleScreenshot,
        mismatchPercentage: 0,
      },
      {
        properties: { key: '/bar', browser: 'Chrome', viewportWidth: 1024 },
        before: exampleScreenshot,
        after: null,
        diff: null,
        mismatchPercentage: 100,
      },
      {
        properties: { key: '/baz', browser: 'Chrome', viewportWidth: 1024 },
        before: exampleScreenshot,
        after: exampleScreenshot,
        diff: exampleScreenshot,
        mismatchPercentage: 50,
      },
    ],
  };

  await websiteReporter.report(report);

  expect(
    await fs.readFile(path.join(outputDirectory, 'index.html'), 'utf8'),
  ).toMatchSnapshot();
});
