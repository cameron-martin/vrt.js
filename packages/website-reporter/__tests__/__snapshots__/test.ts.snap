// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
"<!DOCTYPE html>
<html>
    <head>
        <title>Vrt.js Report</title>
        <style>
            .screenshots {
                display: flex;
            }

            .screenshot {
                width: calc(100% / 3);
            }
        </style>
    </head>
    <body>
        
            <details>
                <summary>{&#34;key&#34;:&#34;/foo&#34;,&#34;browser&#34;:&#34;Chrome&#34;,&#34;viewportWidth&#34;:1024}</summary>
                <div class=\\"screenshots\\">
                    <div class=\\"screenshot\\">
                        <h2>Before</h2>
                        <a href=\\"screenshots/1.png\\"><img src=\\"screenshots/1.png\\"></a>
                    </div>
                    <div class=\\"screenshot\\">
                        <h2>Diff</h2>
                        <a href=\\"screenshots/3.png\\"><img src=\\"screenshots/3.png\\"></a>
                    </div>
                    <div class=\\"screenshot\\">
                        <h2>After</h2>
                        <a href=\\"screenshots/2.png\\"><img src=\\"screenshots/2.png\\"></a>
                    </div>
                </div>
            </details>
        
            <details>
                <summary>{&#34;key&#34;:&#34;/bar&#34;,&#34;browser&#34;:&#34;Chrome&#34;,&#34;viewportWidth&#34;:1024}</summary>
                <div class=\\"screenshots\\">
                    <div class=\\"screenshot\\">
                        <h2>Before</h2>
                        <a href=\\"screenshots/4.png\\"><img src=\\"screenshots/4.png\\"></a>
                    </div>
                    <div class=\\"screenshot\\">
                        <h2>Diff</h2>
                        <a href=\\"screenshots/6.png\\"><img src=\\"screenshots/6.png\\"></a>
                    </div>
                    <div class=\\"screenshot\\">
                        <h2>After</h2>
                        <a href=\\"screenshots/5.png\\"><img src=\\"screenshots/5.png\\"></a>
                    </div>
                </div>
            </details>
        
            <details>
                <summary>{&#34;key&#34;:&#34;/baz&#34;,&#34;browser&#34;:&#34;Chrome&#34;,&#34;viewportWidth&#34;:1024}</summary>
                <div class=\\"screenshots\\">
                    <div class=\\"screenshot\\">
                        <h2>Before</h2>
                        <a href=\\"screenshots/7.png\\"><img src=\\"screenshots/7.png\\"></a>
                    </div>
                    <div class=\\"screenshot\\">
                        <h2>Diff</h2>
                        <a href=\\"screenshots/9.png\\"><img src=\\"screenshots/9.png\\"></a>
                    </div>
                    <div class=\\"screenshot\\">
                        <h2>After</h2>
                        <a href=\\"screenshots/8.png\\"><img src=\\"screenshots/8.png\\"></a>
                    </div>
                </div>
            </details>
        
    </body>
</html>"
`;
